File: internal/engine/select.go
package engine

import (
	"sort"

	"github.com/Cloudbird-Software/mutual/internal/domain"
)

// SelectBudgets 是候选对选择的预算约束（config["budgets"]）。
// nil 指针字段 = 不设上限（与 Python None 语义一致）。
type SelectBudgets struct {
	// PerProfileCap: 每用户最多入选对数
	// （max_n_llm_evaluations_per_profile，member 与 pool 侧同权）。
	PerProfileCap *int
	// GlobalCap: 入选对总数上限（max_pair_llm_calls，A-9：
	// 语义 = 入选 pair 总数上限，每 pair 至少一次 LLM 调用）。
	GlobalCap *int
}

// pairCandidate 是单个用户视角的候选（partner + fused 分数）。
type pairCandidate struct {
	other domain.UserID
	score float64
}

// SelectPairs 从相似度矩阵贪心轮转选择进入 LLM 精排的候选对
// （select 阶段，纯函数）。
//
// 约束（spec/05-boundaries.md §8、spec/02-stages.md §5）：
//   - per-profile cap：每个用户（双侧同权）最多入选 cap 对；
//   - global cap：总对数 ≤ GlobalCap；
//   - novelty：excludedPairs 中的 pair 不参与选择；
//   - 只保留正相似度对（fused 分数 ≤ 0 不选）；
//   - M×N 模式下重叠用户的自配对 (u,u) 排除。
//
// 轮转顺序与平局裁决（spec 沉默 A-7）：每轮按 source_ids 顺序轮流
// 为每个用户取其当前最优候选；分数平局取字典序较小的 partner；
// 返回列表按选择顺序排列。选择依据 = fused_matrix（A-8）。
func SelectPairs(similarity *domain.SimilarityResult, budgets SelectBudgets, excludedPairs map[domain.PairID]bool) []domain.CandidatePair {
	srcIDs := similarity.SourceIDs
	tgtIDs := similarity.TargetIDs
	fused := similarity.FusedMatrix
	square := similarity.IsSquare()

	// 每个源侧用户的候选列表（正相似度、非排除、非自配对）。
	candidates := map[domain.UserID][]pairCandidate{}
	for _, uid := range srcIDs {
		candidates[uid] = nil
	}
	for i, uid := range srcIDs {
		for j, other := range tgtIDs {
			if other == uid {
				continue
			}
			var score float64
			if square && i > j {
				score = fused[j][i] // 上三角镜像（对称 legacy 路径）
			} else {
				score = fused[i][j]
			}
			if score <= 0 {
				continue
			}
			if excludedPairs[domain.StablePairID(uid, other)] {
				continue
			}
			candidates[uid] = append(candidates[uid], pairCandidate{other: other, score: score})
		}
	}
	// 分数降序；平局取字典序较小 partner（A-7）。稳定排序保证确定性。
	for uid := range candidates {
		list := candidates[uid]
		sort.SliceStable(list, func(a, b int) bool {
			if list[a].score != list[b].score {
				return list[a].score > list[b].score
			}
			return list[a].other < list[b].other
		})
		candidates[uid] = list
	}

	counts := map[domain.UserID]int{}
	for _, uid := range srcIDs {
		counts[uid] = 0
	}
	for _, uid := range tgtIDs {
		counts[uid] = 0
	}
	atCap := func(uid domain.UserID) bool {
		return budgets.PerProfileCap != nil && counts[uid] >= *budgets.PerProfileCap
	}

	chosen := map[domain.PairID]bool{}
	var result []domain.CandidatePair
	total := 0

	for budgets.GlobalCap == nil || total < *budgets.GlobalCap {
		progressed := false
		for _, uid := range srcIDs {
			if budgets.GlobalCap != nil && total >= *budgets.GlobalCap {
				break
			}
			if atCap(uid) {
				continue
			}
			for _, c := range candidates[uid] {
				if atCap(c.other) {
					continue
				}
				pid := domain.StablePairID(uid, c.other)
				if chosen[pid] {
					continue
				}
				result = append(result, domain.NewCandidatePair(uid, c.other, c.score))
				chosen[pid] = true
				counts[uid]++
				counts[c.other]++
				total++
				progressed = true
				break
			}
		}
		if !progressed {
			break
		}
	}
	return result
}

File: src/mutual/select.py
"""Mutual — 候选对选择。

对应 docs/engineering-plan.md §3.7、spec/02-stages.md §5。

从相似度矩阵中贪心轮转选择进入 LLM 精排的候选对。
``select_pairs`` 是纯函数，无副作用。

边界（spec/05-boundaries.md §8、spec/02-stages.md §5）：
- per-profile cap：每个用户（member 与 pool 侧同权）最多入选
  ``budgets.max_n_llm_evaluations_per_profile`` 对。
- global cap：总入选对数 ≤ ``budgets.max_pair_llm_calls``。
- novelty：``excluded_pairs``（adapter 按 novelty 窗口构建）中的 pair
  不参与选择。
- 只保留正相似度对（``fused_matrix`` 分数 ≤ 0 的不选）。

spec 沉默：
- A-7：轮转顺序与平局裁决未在 spec 中定义。实现：每轮按
  ``source_ids`` 顺序轮流为每个用户取其当前最优候选；分数平局取
  字典序较小的 partner；返回列表按选择顺序排列。
- A-8：选择依据用 ``fused_matrix``（dir 的最终融合视图）。
- A-9：``max_pair_llm_calls`` 语义 = 入选 pair 总数上限（每 pair 至少
  一次 LLM 调用）。
- A-10：M×N 模式下 source 与 target 重叠用户的自配对（u,u）排除；
  pool 侧用户同样计入 per-profile cap。
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional, Set

from .schemas import CandidatePair, SimilarityResult, stable_pair_id


def select_pairs(
    similarity: SimilarityResult,
    budgets: Dict[str, Any],
    excluded_pairs: Optional[Set[str]] = None,
) -> List[CandidatePair]:
    """贪心轮转选择候选对（per-profile cap + global cap + novelty 排除）。

    Args:
        similarity: 相似度结果（选择依据 = ``fused_matrix``，A-8）。
        budgets: 预算 dict，读取 ``max_n_llm_evaluations_per_profile``
            （per-profile cap）与 ``max_pair_llm_calls``（global cap，A-9）；
            缺省/None = 不设上限。
        excluded_pairs: 排除的 pair_id 集合（novelty 窗口内已暴露的对，
            spec/05-boundaries.md §8）。

    Returns:
        ``list[CandidatePair]``，按贪心选择顺序排列（A-7）；
        ``user1 < user2``，``pair_id = stable_pair_id(user1, user2)``。
    """
    excluded: Set[str] = set(excluded_pairs) if excluded_pairs else set()
    per_cap: Optional[int] = budgets.get("max_n_llm_evaluations_per_profile")
    global_cap: Optional[int] = budgets.get("max_pair_llm_calls")

    src_ids = list(similarity.source_ids)
    tgt_ids = list(similarity.target_ids)
    fused = similarity.fused_matrix
    square = similarity.is_square

    # 每个源侧用户的候选列表（正相似度、非排除、非自配对）。
    # square 模式 fused 对称（legacy 路径），每个用户考虑全部 partner
    # （上三角值镜像），pair 去重由 chosen 集合保证。
    candidates: Dict[str, List[Any]] = {uid: [] for uid in src_ids}
    for i, uid in enumerate(src_ids):
        for j, other in enumerate(tgt_ids):
            if other == uid:
                continue
            if square and i > j:
                score = float(fused[j, i])  # 上三角镜像（对称）
            else:
                score = float(fused[i, j])
            if score <= 0.0:
                continue
            if stable_pair_id(uid, other) in excluded:
                continue
            candidates[uid].append((other, score))
    for uid in candidates:
        # 分数降序；平局取字典序较小的 partner（A-7）。
        candidates[uid].sort(key=lambda t: (-t[1], t[0]))

    counts: Dict[str, int] = {uid: 0 for uid in set(src_ids) | set(tgt_ids)}
    chosen: Set[str] = set()
    result: List[CandidatePair] = []
    total = 0

    def _at_cap(uid: str) -> bool:
        return per_cap is not None and counts[uid] >= per_cap

    while global_cap is None or total < global_cap:
        progressed = False
        for uid in src_ids:
            if global_cap is not None and total >= global_cap:
                break
            if _at_cap(uid):
                continue
            pick = None
            for other, score in candidates[uid]:
                if _at_cap(other):
                    continue
                pid = stable_pair_id(uid, other)
                if pid in chosen:
                    continue
                pick = (other, score, pid)
                break
            if pick is None:
                continue
            other, score, pid = pick
            result.append(CandidatePair.create(uid, other, score))
            chosen.add(pid)
            counts[uid] += 1
            counts[other] += 1
            total += 1
            progressed = True
        if not progressed:
            break
    return result
